aip/general/0236.md

```diff
@@ -0,0 +1,351 @@
+
+1. The user creates an experiment containing a new policy configuration
+   intended to replace the live policy.
+2. The user uses the "preview" method to start generating logs which compare
+   the live and experiment policy evaluations against live traffic.
+3. The user inspects the logs to determine whether the experiment has the
+   intended result.
+4. The user uses the "commit" method to promote the experiment to live.
+
+## Guidance
+
+### Non-goals
+
+This proposal is for a safety mechanism for policy rollouts only. Safe rollouts
+for non-policy resources are not in scope.
+
+### Experiments
+
+A new configuration of a policy to be previewed is stored in a nested
+collection under the policy called experiments.
+
+A hypothetical policy resource called, `Policy`, is used throughout. It has the
+following resource name pattern:
+
+`projects/{project}/locations/{location}/policies/{policy}`
+
+The experimental versions of the resource used for previewing or other safe
+rollout practices are represented as a nested collection under `Policy` using a
+new resource type. The resource type **must** follow the naming convention
+*RegularResourceType*`Experiment`. 
+
+The following pattern is used for the experiment collection:
+
+`projects/{project}/locations/{location}/policies/{policy}/experiments/{experiment}`
+
+A proto used to represent an experiment **must** contain the following:
+
+  1. The required top-level fields for a resource, like `name` and `etag`
+  2. The policy message that is being tested itself
+  3. The field, `preview_metadata`, which contains metadata specific to
+     previewing the experiment.
+
+```proto
+message PolicyExperiment {
+
+  // The resource name of the PolicyExperiment.
+  string name = 1;
+ 
+  // The policy experiment. This Policy will be used to preview the effects of
+  // the change but will not affect life traffic.
+  Policy policy = 2;
+
+  // The metadata associated with this policy experiment.
+  PreviewMetadata preview_metadata = 3 [(google.api.field_behavior) = OUTPUT_ONLY];
+}
+```
+
+- The experiment proto **must** have a top-level field with the same type as the
+  live policy. 
+  - It **must** be named as the live resource type. For example, if the
+    experiment is for FirewallPolicy, then this field **must** be named
+    `firewall_policy`.
+  - The name inside the embedded `policy` message **must** be the name of the
+    live policy. 
+- When the user is ready to promote an experiment, they **must** copy the 
+  `policy` message into the live policy and delete the experiment. This can be
+  done manually or via a "commit" custom method.
+- A product **may** support multiple experiments concurrently being previewed 
+  for a single live policy. 
+  - Each experiment **must** generate logs comparing its results to the live
+    policy.
+  - The number of experimental configurations for a given live policy **may** be
+    capped at a certain number.
+    - The cap **must** be documented.
```

Comment:
One sub-bullet of this depth is a bit unconventional. Would you mind picking up the trailing bullet and just say "and the capy **must* be documented" in the preceding bullet?